Runtime/Components/CinemachineSplineDolly.cs

```diff
@@ -0,0 +1,368 @@
+            }
+
+            // Get the new ideal spline base position
+            if (m_AutoDolly.m_Enabled && FollowTarget != null)
+            {
+                // convert follow target into spline local space, because SplineUtility works in spline local space
+                SplineUtility.GetNearestPoint(spline, 
+                    m_Spline.transform.InverseTransformPoint(FollowTargetPosition), out _, out normalizedSplinePosition, 
+                    m_AutoDolly.m_SearchResolution, m_AutoDolly.m_SearchIteration);
+                m_SplinePosition = SplineUtility.ConvertIndexUnit(spline, normalizedSplinePosition, 
+                    PathIndexUnit.Normalized, m_PositionUnits);
+                
+                // Apply the spline position offset
+                normalizedSplinePosition += SplineUtility.ConvertIndexUnit(
+                    spline, m_AutoDolly.m_PositionOffset, m_PositionUnits, PathIndexUnit.Normalized);
+            }
+
+            if (deltaTime >= 0 && VirtualCamera.PreviousStateIsValid)
+            {
+                float maxUnit = 1; // we are always using normalized unit [0-1]
+                {
+                    float prev = m_PreviousNormalizedSplinePosition;
+                    float next = normalizedSplinePosition;
+                    if (spline.Closed && Mathf.Abs(next - prev) > maxUnit / 2)
+                    {
+                        if (next > prev)
+                            prev += maxUnit;
+                        else
+                            prev -= maxUnit;
+                    }
+                    m_PreviousNormalizedSplinePosition = prev;
+                    normalizedSplinePosition = next;
+                }
+
+                // Apply damping in the spline direction
+                float offset = m_PreviousNormalizedSplinePosition - normalizedSplinePosition;
+                offset = Damper.Damp(offset, m_Damping.z, deltaTime);
+                normalizedSplinePosition = m_PreviousNormalizedSplinePosition - offset;
+            }
+            m_PreviousNormalizedSplinePosition = normalizedSplinePosition;
+            m_Spline.Evaluate(normalizedSplinePosition, 
+                out var localPosition, out var localTangent, out var localUp);
+            Vector3 newCameraPos = localPosition;
+            // fix tangent when 0
+            if (!localTangent.Equals(Vector3.zero))
+            {
+                m_CorrectedTangent = localTangent;
+                if (m_CorrectedTangent.Equals(localUp))
+                {
+                    // Find tangent perpendicular to localUp, localUp is always normalized
+                    var newTangent = new Vector3(localUp.z, localUp.z, -localUp.x - localUp.y);
+                    m_CorrectedTangent = !newTangent.Equals(Vector3.zero)
+                        ? newTangent
+                        : new Vector3(-localUp.y - localUp.z, localUp.x, localUp.x);
+                }
+            }
```

Comment:
Can you explain to me what this code is doing?  I think the comment is out of date.  It looks to me like you're going to get a pop when it executes.  Not that it ever will because the Equals test is too strong for Vector3

Reply:
The tangent is always non-zero, expect when it is set to be 0 by the user. That's why I checked for exact equality.

Reply:
it's actually this equals that I was referring to: `if (m_CorrectedTangent.Equals(localUp))`.
In any case I reimplemented this in my visualization branch, but there are still issues